src/swap/useSwapQuote.ts

```diff
@@ -67,42 +74,132 @@ function createBaseApproveTransaction(
       ? BigInt(new BigNumber(buyAmount).times(guaranteedPrice).toFixed(0, 0))
       : BigInt(sellAmount)
 
-  const data = encodeFunctionData({
-    abi: erc20.abi,
-    functionName: 'approve',
-    args: [allowanceTarget as Address, amountToApprove],
-  })
-
-  const approveTx: TransactionRequestCIP42 & { amountToApprove: bigint } = {
-    from: from as Address,
-    to: sellTokenAddress as Address,
-    data,
-    type: 'cip42',
-    amountToApprove,
+  // Approve transaction if the sell token is ERC-20
+  if (allowanceTarget !== zeroAddress && fromToken.address) {
+    const data = encodeFunctionData({
+      abi: erc20.abi,
+      functionName: 'approve',
+      args: [allowanceTarget as Address, amountToApprove],
+    })
+
+    const approveTx: TransactionRequestCIP42 = {
+      from: from as Address,
+      to: fromToken.address as Address,
+      data,
+    }
+    baseTransactions.push(approveTx)
   }
 
-  return approveTx
-}
-
-function createBaseSwapTransaction(unvalidatedSwapTransaction: SwapTransaction) {
-  const { from, to, value, data, gas } = unvalidatedSwapTransaction
-
   const swapTx: TransactionRequestCIP42 & { gas: bigint } = {
     from: from as Address,
     to: to as Address,
     value: BigInt(value ?? 0),
     data: data as Hex,
     // This may not be entirely accurate for now
+    // without the approval transaction being executed first.
     // See https://www.notion.so/valora-inc/Fee-currency-selection-logic-4c207244893748bd85e23b754334f42d?pvs=4#8b7c27d31ebf4fca981f81e9411f86ee
-    // We'll try to improve this in our API
+    // We control this from our API.
     gas: BigInt(gas),
-    type: 'cip42',
   }
+  baseTransactions.push(swapTx)
+
+  return {
+    amountToApprove,
+    baseTransactions,
+  }
+}
 
-  return swapTx
+function getFeeCurrencyAddress(feeCurrency: TokenBalance) {
+  return !feeCurrency.isNative ? (feeCurrency.address as Address) : undefined
 }
 
-function getMaxGasCost(txs: TransactionRequestCIP42[]) {
+async function tryEstimateTransactions(
+  baseTransactions: TransactionRequestCIP42[],
+  feeCurrency: TokenBalance
+) {
+  const transactions: TransactionRequestCIP42[] = []
+
+  const feeCurrencyAddress = getFeeCurrencyAddress(feeCurrency)
+  const { maxFeePerGas, maxPriorityFeePerGas } = await estimateFeesPerGas(
+    publicClient.celo,
+    feeCurrencyAddress
+  )
+
+  for (const [index, baseTx] of baseTransactions.entries()) {
+    // We can only truly estimate the gas for the first transaction
+    if (index === 0) {
+      const tx = await tryEstimateTransaction(
+        baseTx,
+        feeCurrency,
+        maxFeePerGas,
+        maxPriorityFeePerGas
+      )
+      if (!tx) {
+        return null
+      }
+      transactions.push(tx)
+    } else {
+      if (!baseTx.gas) {
+        throw new Error(
+          'When multiple transactions are provided, all transactions but the first one must have a gas value'
```

Comment:
sorry remind me why the first transaction doesn't need to have a gas value?

Reply:
Because we can estimate it ourselves and be more accurate.